lib/brod_client.ex

```diff
@@ -172,7 +172,7 @@ defmodule BroadwayKafka.BrodClient do
   end
 
   defp validate_option(:hosts, value) do
-    if Keyword.keyword?(value) do
+    if supported_keyword?(value) do
```

Comment:
Hi @gabrielgiordan, this looks good! Although I wouldn't call it a keyword. It is simply a list of two element tuples. So I would change this function name to `supported_hosts` to validate hosts specifically. Similar can be done to `supported_topics`. What is the format accepted for topics?

Reply:
Hi @josevalim, thanks for your review! I agree that we should rename `supported_keyword` to something else since it isn't exactly a keyword list, the difference is that it accepts a string as the key. The format of topics is the same as the hosts' format, only that it also accepts a list of strings, maybe we should use the same function to validate both the topics and the hosts?

The topics' format can be: `["topic_1", "topic_2"]` or `[topic_1: 1, topic_2: 2]` or `[{"topic_1", 1}, {"topic_2", 2}]`

While the hosts' format can be: `[host: 9092]` or `[{"host", 9092}]`

What do you think of something like this:

```elixir
  defp supported_tuple_list?([{key, _value} | rest]) when is_binary(key) or is_atom(key),
    do: supported_hosts?(rest)
  defp supported_tuple_list?([]), do: true
  defp supported_tuple_list?(_other), do: false

  defp supported_hosts?(term), do: supported_tuple_list?(term)

  defp supported_topics?(term), do:
    supported_tuple_list?(term) or (is_list(term) and Enum.all?(term, &is_binary/1))
```

Reply:
Yeah, this sounds good to me. :) Although in the long term I would probably move the hosts format to `"host:123"`. But that's a separate story.

Btw, what is the number given to a topic? The number of partitions?

Reply:
Your question made me debug the topic/partition option, I think we currently support only a list of topics:

```elixir
09:48:54.434 [info]  [supervisor: {:local, :brod_sup}, started: [pid: #PID<0.241.0>, id: Consumer.Broadway.Producer_0.Client, mfargs: {:brod_client, :start_link, [["##############.aws.confluent.cloud": 9092], Consumer.Broadway.Producer_0.Client, [sasl: {:plain, "#################", #Function<11.113019958/0 in :brod_utils.init_sasl_opt/1>}, ssl: [cacertfile: "./priv/ssl/server.crt", keyfile: "./priv/ssl/server.key", certfile: "./priv/ssl/server.crt"]]]}, restart_type: {:permanent, 10}, shutdown: 5000, child_type: :worker]]
Interactive Elixir (1.9.4) - press Ctrl+C to exit (type h() ENTER for help)
iex(1)> 
09:48:56.857 [info]  Group member (test,coor=#PID<0.245.0>,cb=#PID<0.240.0>,generation=0):
Leaving group, reason: {{:function_clause, [:topics, :protocol_metadata, :group_protocols, {:join_group, 0}], [group_id: "test", session_timeout: 30000, member_id: "", protocol_type: "consumer", group_protocols: [[protocol_name: :roundrobin_v2, protocol_metadata: [version: 0, topics: [test: 0], user_data: ""]]]]}, [{:kpro_lib, :encode, [:string, {:test, 0}], []}, {:kpro_req_lib, :enc_struct_field, 3, [file: 'src/kpro_req_lib.erl', line: 398]}, {:kpro_req_lib, :"-enc_struct_field/3-lc$^0/1-0-", 3, [file: 'src/kpro_req_lib.erl', line: 389]}, {:kpro_req_lib, :enc_struct_field, 3, [file: 'src/kpro_req_lib.erl', line: 389]}, {:kpro_req_lib, :enc_struct, 3, [file: 'src/kpro_req_lib.erl', line: 379]}, {:kpro_req_lib, :enc_struct, 3, [file: 'src/kpro_req_lib.erl', line: 380]}, {:kpro_req_lib, :translate, 2, [file: 'src/kpro_req_lib.erl', line: 410]}, {:kpro_req_lib, :enc_struct, 3, [file: 'src/kpro_req_lib.erl', line: 378]}]}

 
09:48:56.863 [error] GenServer #PID<0.245.0> terminating
** (stop) {:function_clause, [:topics, :protocol_metadata, :group_protocols, {:join_group, 0}], [group_id: "test", session_timeout: 30000, member_id: "", protocol_type: "consumer", group_protocols: [[protocol_name: :roundrobin_v2, protocol_metadata: [version: 0, topics: [test: 0], user_data: ""]]]]}
    (kafka_protocol) :kpro_lib.encode(:string, {:test, 0})
    (kafka_protocol) src/kpro_req_lib.erl:398: :kpro_req_lib.enc_struct_field/3
    (kafka_protocol) src/kpro_req_lib.erl:389: :kpro_req_lib."-enc_struct_field/3-lc$^0/1-0-"/3
    (kafka_protocol) src/kpro_req_lib.erl:389: :kpro_req_lib.enc_struct_field/3
    (kafka_protocol) src/kpro_req_lib.erl:379: :kpro_req_lib.enc_struct/3
    (kafka_protocol) src/kpro_req_lib.erl:380: :kpro_req_lib.enc_struct/3
    (kafka_protocol) src/kpro_req_lib.erl:410: :kpro_req_lib.translate/2
    (kafka_protocol) src/kpro_req_lib.erl:378: :kpro_req_lib.enc_struct/3
Last message: {:lo_cmd_stabilize, 0, :undefined}
State: {:state, Consumer.Broadway.Producer_0.Client, "test", "", :undefined, 0, [test: 0], :undefined, :undefined, [], false, #PID<0.240.0>, BroadwayKafka.Producer, [], :undefined, :roundrobin_v2, 30, 5, 5, 2, :undefined, :commit_to_kafka_v2, 1, :roundrobin_v2} 
```

Given that, I will leave the topics as it is in this PR and open an issue to resolve it later.

Reply:
On the new format "host:123", good idea! In the future maybe I'll open a PR with this :v: